File: retos-ts/20_faltantes_y_duplicados.ts
// Resultado 5 ⭐
function fixGiftList(received: string[], expected: string[]): { missing: Record<string, number>, extra: Record<string, number> } {
    const missing = {}
  const extra = {}
  const objReceived = {}
  const objExpected = {}

  for(const gift of received){
    objReceived[gift] 
      ? objReceived[gift]++
      : objReceived[gift] = 1
  }

  if(expected.length === 0){
    for(const gift of Object.keys(objReceived)){
      extra[gift] = objReceived[gift]
    }

    return { missing, extra } 
  }

  for(const gift of expected){
    objExpected[gift] 
      ? objExpected[gift]++
      : objExpected[gift] = 1
  }

  for(const gift of Object.keys(objExpected)){
    if(!objReceived[gift]){
      missing[gift] = objExpected[gift] 
    }else{
      const giftExp = objExpected[gift]
      const giftRec = objReceived[gift]

      if(giftExp < giftRec){
        extra[gift] = giftRec - giftExp
      }

      if(giftExp > giftRec){
        missing[gift] = giftExp - giftRec
      }
    }
  }

  for(const gift of Object.keys(objReceived)){
    if(!objExpected[gift]){
      extra[gift] = objReceived[gift]
    }
  }

  return { missing, extra }
}
    

File: retos-python/20_faltantes_y_duplicados.py
# Resultado 5 ⭐
def fix_gift_list(received, expected):
    missing = {}
    extra = {}
    obj_received = {}
    obj_expected = {}

    for gift in received:
        if gift in obj_received:
            obj_received[gift] += 1
        else:
            obj_received[gift] = 1

    if not expected:
        for gift, count in obj_received.items():
            extra[gift] = count
        return {"missing": missing, "extra": extra}

    for gift in expected:
        if gift in obj_expected:
            obj_expected[gift] += 1
        else:
            obj_expected[gift] = 1

    for gift, count_expected in obj_expected.items():
        count_received = obj_received.get(gift, 0)

        if count_received == 0:
            missing[gift] = count_expected
        elif count_expected > count_received:
            missing[gift] = count_expected - count_received
        elif count_expected < count_received:
            extra[gift] = count_received - count_expected

    for gift, count_received in obj_received.items():
        if gift not in obj_expected:
            extra[gift] = count_received

    return {"missing": missing, "extra": extra}